src/eachobs.jl

```diff
@@ -42,18 +50,80 @@ for (x, y) in eachobs((X, Y))
 end
 ```
 """
-function eachobs(data; buffer = false, batchsize::Int = -1, partial::Bool =true)
+function eachobs(
+        data;
+        buffer = false,
+        parallel = false,
+        shuffle = false,
+        batchsize::Int = -1,
+        partial::Bool = true,
+        executor = _default_executor())
     if batchsize > 0
         data = BatchView(data; batchsize, partial)
     end
-    if buffer === false 
-        gen = (getobs(data, i) for i in 1:numobs(data))
+
+    iter = if parallel
+        eachobsparallel(data; buffer, executor)
     else
-        if buffer === true && numobs(data) > 0
-            buffer = getobs(data, 1)
+        if buffer === false
+            EachObs(data)
+        elseif buffer === true
+            EachObsBuffer(data)
+        else
+            EachObsBuffer(data, buffer)
         end
-        gen = (getobs!(buffer, data, i) for i in 1:numobs(data))
     end
-    return gen
+
+    if shuffle
+        iter = ReshuffleIter(iter)
+    end
+    return iter
+end
+
+
+# Internal
+
+"""
+    EachObs(data)
+
+Create an iterator over observations in data container `data`.
+
+This is an internal function. Use `eachobs(data)` instead.
+"""
+struct EachObs{T}
+    data::T
 end
+Base.length(iter::EachObs) = numobs(iter.data)
+Base.eltype(iter::EachObs) = eltype(iter.data)
 
+function Base.iterate(iter::EachObs, i::Int = 1)
+    i > numobs(iter) && return nothing
+    return getobs(iter.data, i), i+1
+end
+
+
+"""
+    EachObsBuffer(data)
+
+Create a buffered iterator over observations in data container `data`.
+Buffering only works if [`getobs!`](@ref) is implemented for `data`.
+
+This is an internal function. Use `eachobs(data, buffer = true)` instead.
+"""
+struct EachObsBuffer{T, B}
+    data::T
+    buffer::B
+end
+EachObsBuffer(data) = EachObsBuffer(data, getobs(data, 1))
+Base.length(iter::EachObsBuffer) = numobs(iter.data)
+Base.eltype(iter::EachObsBuffer) = eltype(iter.data)
+
+function Base.iterate(iter::EachObsBuffer)
+    obs = getobs!(iter.buffer, iter.data, 1)
+    return obs, (obs, 2)
+end
+
+function Base.iterate(iter::EachObsBuffer, (buffer, i))
```

Comment:
```suggestion
function Base.iterate(iter::EachObsBuffer, (buffer, i) = (nothing, 1))
```
and remove `iterate(iter)` above

Reply:
Shouldn't this be 
```suggestion
function Base.iterate(iter::EachObsBuffer, (buffer, i) = (iter.buffer, 1))
```
then? So that the stored buffer is used.

Reply:
yes